flink-runtime/src/main/java/org/apache/flink/runtime/state/UserFacingTemporalListState.java

```diff
@@ -33,18 +36,45 @@
     private final Iterable<T> emptyState = Collections.emptyList();
 
     UserFacingTemporalListState(ListState<T> originalState) {
-
         this.originalState = originalState;
         this.internalState = (InternalListState) originalState;
     }
 
     @Override
-    public Iterable<T> get(long time) throws Exception {
+    public Iterable<T> getAt(long time) throws Exception {
         internalState.setCurrentNamespace(time);
         Iterable<T> original = originalState.get();
         return original != null ? original : emptyState;
     }
 
+    @Override
+    public TemporalValue<Iterable<T>> getAtOrBefore(long time) throws Exception {
+        TemporalNavigator navigator = Preconditions.checkNotNull(internalState.getNavigator());
+        Long found = navigator.atOrBefore(time);
+        if (found == null) {
+            return new TemporalValue<>(Long.MIN_VALUE, emptyState);
+        }
+
+        internalState.setCurrentNamespace(found);
+        Iterable<T> original = originalState.get();
+        Iterable<T> iter = original != null ? original : emptyState;
+        return new TemporalValue<>(found, iter);
+    }
+
+    @Override
+    public TemporalValue<Iterable<T>> getAtOrAfter(long time) throws Exception {
+        TemporalNavigator navigator = Preconditions.checkNotNull(internalState.getNavigator());
+        Long found = navigator.atOrAfter(time);
+        if (found == null) {
+            return new TemporalValue<>(Long.MIN_VALUE, emptyState);
+        }
+
+        internalState.setCurrentNamespace(found);
+        Iterable<T> original = originalState.get();
+        Iterable<T> iter = original != null ? original : emptyState;
```

Comment:
I think, I'd like to extend the current functionality to retrieve key+value at once (from the navigator) - I'll prepare a separate PR for this and we can see about the practical difference)